### PR TITLE
Create instagram_crawler.py

### PR DESCRIPTION
**Pull request by yogeshwaran01**
### **Describe your change:**
Crawl the Instagram users information

* [X] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [X] This pull request is all my own work -- I have not plagiarized.
* [X] I know that pull requests will not be merged if they fail the automated tests.
* [X] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [X] All new Python files are placed inside an existing directory.
* [X] All filenames are in all lowercase characters with no spaces or dashes.
* [X] All functions and variable names follow Python naming conventions.
* [X] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [X] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [X] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [X] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: web_programming/instagram_crawler.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*-
+import requests
+from bs4 import BeautifulSoup
+import json
+
+headers = {
+    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36"
```
Comment (cclauss) [inline web_programming/instagram_crawler.py:8]:
```suggestion
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_5) "
    "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36"
```

File: web_programming/instagram_crawler.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+        html = requests.get(self.url, headers=headers)
+        soup = BeautifulSoup(html.text, "html.parser")
+        try:
+            return html_1(soup)
+        except json.decoder.JSONDecodeError:
+            return html_2(soup)
+
+    @property
+    def no_of_followers(self) -> int:
```
Comment (cclauss) [inline web_programming/instagram_crawler.py:44]:
```suggestion
    def number_of_followers(self) -> int:
```

Comment (cclauss) [reply web_programming/instagram_crawler.py:45]: or `follower_count`

File: web_programming/instagram_crawler.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    def no_of_posts(self) -> int:
+        """
+        return number of posts
+        """
+
+        info = self.get_json()
+        return info["edge_owner_to_timeline_media"]["count"]
+
+    @property
+    def get_biography(self) -> str:
```
Comment (cclauss) [inline web_programming/instagram_crawler.py:72]:
```suggestion
    def biography(self) -> str:
```

File: web_programming/instagram_crawler.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        """
+        return biography of user
+        """
+
+        info = self.get_json()
+        return info["biography"]
+
+    @property
+    def get_fullname(self) -> str:
```
Comment (cclauss) [inline web_programming/instagram_crawler.py:81]:
```suggestion
    def fullname(self) -> str:
```

File: web_programming/instagram_crawler.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    def get_username(self) -> str:
+        """
+        return the username of the user
+        """
+
+        info = self.get_json()
+        return info["username"]
+
+    @property
+    def get_profile_pic(self) -> str:
```
Comment (cclauss) [inline web_programming/instagram_crawler.py:99]: Is the picture really a str?

Comment (yogeshwaran01) [reply web_programming/instagram_crawler.py:99]: yes, It is a link

Comment (cclauss) [reply web_programming/instagram_crawler.py:99]: def profile_picture_url(self) -> str:

File: web_programming/instagram_crawler.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/python3
```
Comment (cclauss) [inline web_programming/instagram_crawler.py:1]:
```suggestion
#!/usr/bin/env python3
```

File: web_programming/instagram_crawler.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+# -*- coding: utf-8 -*-
```
Comment (cclauss) [inline web_programming/instagram_crawler.py:2]:
```suggestion
```
That line only needed to exist on Python 2 which died on 1/1/2020.

File: web_programming/instagram_crawler.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+class InstagramUser:
+    """
+    Class Instagram crawl instagram user information
+    """
+
+    def __init__(self, username):
+        self.username = username
+        self.url = f"https://www.instagram.com/{username}/"
+
+    def get_json(self):
```
Comment (cclauss) [inline web_programming/instagram_crawler.py:31]:
```suggestion
    def get_json(self) -> dict:
```

**Review by cclauss — CHANGES_REQUESTED** [review]

**Review by yogeshwaran01 — COMMENTED** [review]
done

**Comment by yogeshwaran01** [comment]
Requested changes are commited.

File: web_programming/instagram_crawler.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+class InstagramUser:
+    """
+    Class Instagram crawl instagram user information
+    """
+
+    def __init__(self, username):
+        self.username = username
+        self.url = f"https://www.instagram.com/{username}/"
+
+    def get_json(self) -> dict:
```
Comment (yogeshwaran01) [inline web_programming/instagram_crawler.py:31]: requested changes have done

File: web_programming/instagram_crawler.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    def get_username(self) -> str:
+        """
+        return the username of the user
+        """
+
+        info = self.get_json()
+        return info["username"]
+
+    @property
+    def profile_picture_url(self) -> str:
```
Comment (yogeshwaran01) [inline web_programming/instagram_crawler.py:98]: requested changes done

File: web_programming/instagram_crawler.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        """
+        html = requests.get(self.url, headers=headers).text
+        scripts = BeautifulSoup(html, "html.parser").find_all("script")
+        try:
+            return extract_user_profile(scripts[4])
+        except (json.decoder.JSONDecodeError, KeyError):
+            return extract_user_profile(scripts[3])
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}({self.username})"
```
Comment (dhruvmanila) [inline web_programming/instagram_crawler.py:52]:
```suggestion
    def __repr__(self) -> str:
        return f"{self.__class__.__name__}('{self.username}')"
```
Here, the username is going to be a `str` object. `repr` is the string representation that will evaluate to a Python object. A good test to keep in mind while writing the code for `repr` is `obj == eval(repr(obj))`

**Comment by dhruvmanila** [comment]
I think the `user_agent` differs with the platform and browser used.

***From wikipedia:***
`Mozilla/[version] ([system and browser information]) [platform] ([platform details]) [extensions]`

```
"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_5) "
     "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36"
```

Here, you are using a `user_agent` assuming to be on Macintosh and the browsers to be either Chrome or Safari while our Travis tests are run on Linux. I don't have much idea regarding this topic so please do the research and update the `user_agent` accordingly.

**Comment by cclauss** [comment]
In this repo, we have https://pypi.org/project/fake-useragent/ installed.

**Comment by dhruvmanila** [comment]
Please do these small checks locally so that you don't waste your time editing over and over again. 
Thank you for your cooperation 😁

- `black <filename>`
- `isort <filename>`
- `flake8 --ignore=E203,W503 --max-complexity=25 --max-line-length=88 --statistics --count <filename>`
- `mypy --ignore-missing-imports <filename>`
- `pytest --doctest-modules <filename>`

**Comment by cclauss** [comment]
@dhruvmanila Are you interested in creating a [pre-commit hook](https://pre-commit.com/) for us?

https://github.com/internetarchive/openlibrary/blob/master/.pre-commit-config.yaml

**Comment by dhruvmanila** [comment]
Sure, I'll take a look at it. I've never tried one, so it will be a fun challenge!

**Comment by dhruvmanila** [comment]
Hey, @yogeshwaran01, Please run this command locally and see what you're getting as the output:
```console
pytest --doctest-modules instagram_crawler.py
```

**Comment by yogeshwaran01** [comment]
`pytest --doctest-modules instagram_crawler.py`

 It passed the test

`3 passed in 9.11s`

**Comment by yogeshwaran01** [comment]
Wow, It worked! Thank you, @cclauss and @dhruvmanila.

**Comment by dhruvmanila** [comment]
Your welcome! 🎉 😃 
@cclauss Is it ok to have no tests?

**Review by cclauss — APPROVED** [review]
Looks good to me.  The doctest run locally for both @yogeshwaran01 and I but they fail on Travis CI.

**Comment by cclauss** [comment]
There is a gold star for whoever can make the doctests pass under Travis CI.

**Pull request merged**